openhgnn/sampler/MAGNN_sampler.py

```diff
@@ -0,0 +1,195 @@
+import dgl
+import numpy as np
+from openhgnn.models.MAGNN import mp_instance_sampler, mini_mp_instance_sampler
+
+class MAGNN_sampler():
+    '''
+    Description
+    -----------
+    MAGNN sampler based on the implementation of author. The author only implements one sampling 1-layer subgraphs. Here
+    we implement the MAGNN sampler which can sample n-layer subgraphs based on DGL and Pytorch.
+
+    The MAGNN sampler samples n-layer subgraphs based on seed_node of a certain class, metapath instances of seed_nodes
+    which contain all neighbors that might be sampled.
+
+    Users can make use of the MAGNN sampler as a dataset of torch.DataLoader so that we can assign num_workers to
+    concurrently speed up sampling subgraphs.
+
+    Parameters
+    ----------
+    g : dgl.DGLGraph
+        the dgl graph from which we sample subgraphs.
+    mask : list or numpy.ndarray
+        the train/test/test nodes mask, which should be like train mask = [0, 1, 0, 0, 1] indicating that the
+        1th, 4th nodes are training nodes of g. Note that the mask indicates which part of nodes in g will be sampled
+        by MAGNN sampler. e.g. if it's the test mask, the sample will be employed on testing nodes of g to sample
+        subgraphs for testing phase. This param is useful when the graph is too large to be directly validate on.
+    n_layers : int
+        the number of layers of each subgraph.
+    category : any
+        the class of seed_nodes.
+    metapath_list : list
+        the type of all metapaths in graph g. e.g. ['MAM', 'MDM', ...]
```

Comment:
`List of all metapaths to sample, e.g. ...`

Also
* Do they have to have the same starting node type and destination node type?
* How does each element map to the actual metapaths in the graph?  Specifically, if the graph is a custom graph, does feeding in a string to represent a metapath make sense?

Reply:
* Yes the metapaths have to be symmetric so the starting node and the ending node must be the same
* Actually the metapath instances in graph are sampled by `mp_instance_sampler` in MAGNN.py. Before training, we firstly sample all metapath instances based on the whole graph and the given `metapath_list`. Each of elements in `metapath_list` corresponds to a certain type of metapath instances. 
* Moreover, if the graph is not a HGNN, there's no sampled metapath instances. MAGNN cannot be employed on this kind of graph. Because MAGNN is set to run on metapath instances based on given `metapath_list` and sample metapath instances.

Reply:
> Actually the metapath instances in graph are sampled by mp_instance_sampler in MAGNN.py. Before training, we firstly sample all metapath instances based on the whole graph and the given metapath_list. Each of elements in metapath_list corresponds to a certain type of metapath instances.

I understand the semantics.  What I'm asking is more on the API level.  Specifically, what if the node/edge type name is a long string?  I saw that L494-499 is building a dictionary whose keys are the pair of first and last characters.  This doesn't seem right.

Reply:
Or you can specify the list of edge types as metapaths (which is what we have in `dgl.sampling.random_walk`).

Reply:
Actually I think the metapaths have to be specified. Because sometimes the metapaths we need to sampled from the graph are not all combinations of edge types. So if we only specify edge types, we cannot know the actual metapaths.

Reply:
No I meant specifying metapaths as a list of edge types, like:
```python
metapath = ['paper-writtenby-author', 'author-writes-paper']
```

Reply:
Oh yeah I think it's great. I'm thinking about it right now and about to try it.

Reply:
I've fixed this problem in new commit.